Libplanet/Blockchain/Policies/VolatileStagePolicy.cs

```diff
@@ -248,37 +293,73 @@ public long GetNextTxNonce(BlockChain<T> blockChain, Address address)
             return nonce;
         }
 
-        private bool Expired(Transaction<T> transaction) =>
-            transaction.Timestamp + Lifetime < DateTimeOffset.UtcNow;
-
+        /// <summary>
+        /// Retrieves a staged <see cref="Transaction{T}"/> matching given <paramref name="id"/>.
+        /// </summary>
+        /// <param name="blockChain">The <see cref="BlockChain{T}"/> to use
+        /// as a reference when filtering.</param>
+        /// <param name="id">The <see cref="TxId"/> to look up.</param>
+        /// <param name="filtered">Whether to return filtered result or not.</param>
+        /// <returns>
+        /// Returned value is decided according to the following rule:
+        /// <list type="bullet">
+        ///   <item><description>
+        ///     If <paramref name="id"/> is not staged, returns <see langword="null"/>.
+        ///   </description></item>
+        ///   <item><description>
+        ///     If <paramref name="id"/> is staged and expired, returns the found
+        ///     <see cref="Transaction{T}"/> earmarked with <see langword="true"/>.
+        ///     This bypasses filtering.
+        ///   </description></item>
+        ///   <item><description>
+        ///     If <paramref name="id"/> is staged, not expired, and filtered,
+        ///     returns <see langword="null"/>.  This is treating the same as if
+        ///     the <see cref="Transaction{T}"/> was not staged in the first place.
+        ///   </description></item>
+        ///   <item><description>
+        ///     If <paramref name="id"/> is staged, not expired, and not filtered,
+        ///     returns the found <see cref="Transaction{T}"/> earmarked
+        ///     with <see langword="false"/>.
+        ///   </description></item>
+        /// </list>
+        /// </returns>
         /// <remarks>
-        /// It has been intended to avoid recursive lock, hence doesn't hold any synchronous scope.
-        /// Therefore, we should manage the lock from its caller side.
+        /// <para>
+        /// This has been intended to avoid recursive lock, hence doesn't hold any synchronous
+        /// scope.  Therefore, we should manage the lock from its caller side.
+        /// </para>
+        /// <para>
+        /// Earmarking is done to defer removal of the expired <see cref="Transaction{T}"/> so
+        /// the act of inner retrieval can be write-lock-free.  This allows the caller to
+        /// aggregate expired <see cref="Transaction{T}"/>s and remove them in bulk.
+        /// </para>
         /// </remarks>
-        private Transaction<T>? GetInner(BlockChain<T> blockChain, TxId id, bool filtered)
+        private (Transaction<T> Transaction, bool Expired)? GetInner(
+            BlockChain<T> blockChain, TxId id, bool filtered)
         {
             if (_staged.TryGetValue(id, out Transaction<T>? tx) && tx is { })
             {
-                if (Expired(tx) || _ignored.Contains(tx.Id))
```

Comment:
is it okay we don't care about `_ignored` here?

Reply:
`Ignore()` has been changed to manage `_ignored` to be in tandem with `_staged`. The only way for a `TxId` to be marked ignored is through `Ignore()` and when marked, it is always removed from `_staged`.